pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -105,11 +105,11 @@ func (wes WhenExpressions) HaveVariables() bool {
 }
 
 // ReplaceWhenExpressionsVariables interpolates variables, such as Parameters and Results, in
-// the Input and Values.
+// the Input and Values. Returns a new instance of WhenExpressions.
 func (wes WhenExpressions) ReplaceWhenExpressionsVariables(replacements map[string]string) WhenExpressions {
-	replaced := wes
-	for i := range wes {
-		replaced[i] = wes[i].applyReplacements(replacements)
+	var replaced []WhenExpression
```

Comment:
@jerop we might run into that kind of memory issue again in future if we decide to introduce `when` expressions in finally tasks.

Reply:
@bobcatfish This does not fix the underlying issue. Notice, the parameter resolution is successful. `$(params.first-param)` and `$(params.second-param)` are replaced with their respective values in `tasks` but its not reflected in `p`. By separating it out (which is great, I do not object) we are just avoiding calling `ReplaceWhenExpressionsVariables` when `finally` exists. 

```
go test ./pkg/reconciler/pipelinerun/resources/... -tags=unit -run=TestApplyParameters
(string) (len=6) "tasks:"
([]v1beta1.PipelineTask) (len=2 cap=2) {
 (v1beta1.PipelineTask) {
  Name: (string) (len=12) "first-task-1",
  TaskRef: (*v1beta1.TaskRef)(0xc0003ca8d0)({
   Name: (string) (len=10) "first-task",
   Kind: (v1beta1.TaskKind) "",
   APIVersion: (string) ""
  }),
  TaskSpec: (*v1beta1.EmbeddedTask)(<nil>),
  Conditions: ([]v1beta1.PipelineTaskCondition) <nil>,
  WhenExpressions: (v1beta1.WhenExpressions) (len=1 cap=1) {
   (v1beta1.WhenExpression) {
    Input: (string) (len=13) "default-value",
    Operator: (selection.Operator) (len=2) "in",
    Values: ([]string) (len=1 cap=1) {
     (string) (len=12) "second-value"
    }
   }
  },
  Retries: (int) 0,
  RunAfter: ([]string) <nil>,
  Resources: (*v1beta1.PipelineTaskResources)(<nil>),
  Params: ([]v1beta1.Param) <nil>,
  Workspaces: ([]v1beta1.WorkspacePipelineTaskBinding) <nil>,
  Timeout: (*v1.Duration)(<nil>)
 },
 (v1beta1.PipelineTask) {
  Name: (string) (len=12) "final-task-1",
  TaskRef: (*v1beta1.TaskRef)(0xc0003ca900)({
   Name: (string) (len=10) "first-task",
   Kind: (v1beta1.TaskKind) "",
   APIVersion: (string) ""
  }),
  TaskSpec: (*v1beta1.EmbeddedTask)(<nil>),
  Conditions: ([]v1beta1.PipelineTaskCondition) <nil>,
  WhenExpressions: (v1beta1.WhenExpressions) <nil>,
  Retries: (int) 0,
  RunAfter: ([]string) <nil>,
  Resources: (*v1beta1.PipelineTaskResources)(<nil>),
  Params: ([]v1beta1.Param) <nil>,
  Workspaces: ([]v1beta1.WorkspacePipelineTaskBinding) <nil>,
  Timeout: (*v1.Duration)(<nil>)
 }
}
(string) (len=2) "p:"
(*v1beta1.PipelineSpec)(0xc0003b8780)({
 Description: (string) "",
 Resources: ([]v1beta1.PipelineDeclaredResource) <nil>,
 Tasks: ([]v1beta1.PipelineTask) (len=1 cap=1) {
  (v1beta1.PipelineTask) {
   Name: (string) (len=12) "first-task-1",
   TaskRef: (*v1beta1.TaskRef)(0xc0003ca8d0)({
    Name: (string) (len=10) "first-task",
    Kind: (v1beta1.TaskKind) "",
    APIVersion: (string) ""
   }),
   TaskSpec: (*v1beta1.EmbeddedTask)(<nil>),
   Conditions: ([]v1beta1.PipelineTaskCondition) <nil>,
   WhenExpressions: (v1beta1.WhenExpressions) (len=1 cap=1) {
    (v1beta1.WhenExpression) {
     Input: (string) (len=21) "$(params.first-param)",
     Operator: (selection.Operator) (len=2) "in",
     Values: ([]string) (len=1 cap=1) {
      (string) (len=22) "$(params.second-param)"
     }
    }
   },
   Retries: (int) 0,
   RunAfter: ([]string) <nil>,
   Resources: (*v1beta1.PipelineTaskResources)(<nil>),
   Params: ([]v1beta1.Param) <nil>,
   Workspaces: ([]v1beta1.WorkspacePipelineTaskBinding) <nil>,
   Timeout: (*v1.Duration)(<nil>)
  }
 },
 Params: ([]v1beta1.ParamSpec) (len=2 cap=2) {
  (v1beta1.ParamSpec) {
   Name: (string) (len=11) "first-param",
   Type: (v1beta1.ParamType) (len=6) "string",
   Description: (string) "",
   Default: (*v1beta1.ArrayOrString)(0xc0003e8140)({
    Type: (v1beta1.ParamType) (len=6) "string",
    StringVal: (string) (len=13) "default-value",
    ArrayVal: ([]string) <nil>
   })
  },
  (v1beta1.ParamSpec) {
   Name: (string) (len=12) "second-param",
   Type: (v1beta1.ParamType) (len=6) "string",
   Description: (string) "",
   Default: (*v1beta1.ArrayOrString)(<nil>)
  }
 },
 Workspaces: ([]v1beta1.PipelineWorkspaceDeclaration) <nil>,
 Results: ([]v1beta1.PipelineResult) <nil>,
 Finally: ([]v1beta1.PipelineTask) (len=1 cap=1) {
  (v1beta1.PipelineTask) {
   Name: (string) (len=12) "final-task-1",
   TaskRef: (*v1beta1.TaskRef)(0xc0003ca900)({
    Name: (string) (len=10) "first-task",
    Kind: (v1beta1.TaskKind) "",
    APIVersion: (string) ""
   }),
   TaskSpec: (*v1beta1.EmbeddedTask)(<nil>),
   Conditions: ([]v1beta1.PipelineTaskCondition) <nil>,
   WhenExpressions: (v1beta1.WhenExpressions) <nil>,
   Retries: (int) 0,
   RunAfter: ([]string) <nil>,
   Resources: (*v1beta1.PipelineTaskResources)(<nil>),
   Params: ([]v1beta1.Param) <nil>,
   Workspaces: ([]v1beta1.WorkspacePipelineTaskBinding) <nil>,
   Timeout: (*v1.Duration)(<nil>)
  }
 }
})
--- FAIL: TestApplyParameters (0.00s)
    --- FAIL: TestApplyParameters/single_parameter_with_when_expression (0.00s)
        apply_test.go:182: ApplyParameters() got diff (-want, +got):   &v1beta1.PipelineSpec{
              	Description: "",
              	Resources:   nil,
              	Tasks: []v1beta1.PipelineTask{
              		{
              			... // 2 identical fields
              			TaskSpec:   nil,
              			Conditions: nil,
              			WhenExpressions: v1beta1.WhenExpressions{
              				{
            - 					Input:    "default-value",
            + 					Input:    "$(params.first-param)",
              					Operator: "in",
            - 					Values:   []string{"second-value"},
            + 					Values:   []string{"$(params.second-param)"},
              				},
              			},
              			Retries:  0,
              			RunAfter: nil,
              			... // 4 identical fields
              		},
              	},
              	Params:     {{Name: "first-param", Type: "string", Default: &{Type: "string", StringVal: "default-value"}}, {Name: "second-param", Type: "string"}},
              	Workspaces: nil,
              	... // 2 identical fields
              }
FAIL
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/resources	1.599s
FAIL
```

Reply:
Introducing `append` here might take us into inconsistent behavior in future 😞. 

We have been replacing `params` in place (may be change that too 🤔): 

https://github.com/tektoncd/pipeline/blob/81802739f5356f927c4f976a7a3f000fee95b773/pkg/reconciler/pipelinerun/resources/apply.go#L106-L110

Reply:
@pritidesai you were right, using append like this had some unexpected side effects!

I spent a bunch of time trying to understand why the mutations were consistently succeeding, but returning a new when expressions list wasn't and I think I finally understand it.

I created this example to illustrate it: https://play.golang.org/p/Fdnf84zT4L-

I'm not 100% what the best solution is to support when expressions in finally tasks, but i also think we can cover that when we get there.

I think what ended up biting us here is that ApplyReplacements is trying to mix mutating and avoiding mutation (which was my recommendation! apologies @jerop !)

```
p.Tasks[i].WhenExpressions = p.Tasks[i].WhenExpressions.ReplaceWhenExpressionsVariables(replacements)
```

The above statement is subtly trying to do both, b/c it's trying to mutate `p.Tasks[i].WhenExpressions` but unfortunately it's in some cases mutating a copy and in some cases mutating the original

Reply:
I was thinking of this more overnight and I think the other kinda "root cause" here is that when we were using append(), we were then iterating over the resulting slice and mutating it, as if it would mutate the copy "p", but it was almost a coincidence that it did, because we're using structs by value. it just so happened we could mutate the values in "p" because it happened that the copy semantics for slices copied the pointer and still pointed at the slices in "p"

/me gets off soapbox

Reply:
thank you @pritidesai and @bobcatfish!

undid the revert of the mutation and will leave that to be something to be explored in the future

so this pr only separates the replacements in tasks and in finally tasks